media_cdn/dualtoken.py

```diff
@@ -0,0 +1,174 @@
+#
+# Copyright 2022 Google, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# [START  mediacdn_dual_token_sign_token]
+import base64
+import datetime
+import hashlib
+import hmac
+
+import cryptography.hazmat.primitives.asymmetric.ed25519 as ed25519
+
+
+def base64_encoder(value: bytes) -> str:
+    """
+    Returns a base64-encoded string compatible with Media CDN.
+
+    Media CDN uses URL-safe base64 encoding and strips off the padding at the
+    end.
+    """
+    encoded_bytes = base64.urlsafe_b64encode(value)
+    encoded_str = encoded_bytes.decode("utf-8")
+    return encoded_str.rstrip("=")
+
+
+def sign_token(
+    base64_key: bytes,
+    signature_algorithm: str,
+    start_time: datetime.datetime = None,
+    expiration_time: datetime.datetime = None,
+    url_prefix: str = None,
+    full_path: str = None,
+    path_globs: str = None,
+    session_id: str = None,
+    data: str = None,
+    headers: str = None,
+    ip_ranges: str = None,
+) -> str:
+    """Gets the Signed URL Suffix string for the Media CDN' Short token URL requests.
+    One of (`url_prefix`, `full_path`, `path_globs`) must be included in each input.
+    Args:
+        base64_key: Secret key as a base64 encoded string.
+        signature_algorithm: Algorithm can be either `SHA1` or `SHA256` or `Ed25519`.
+        start_time: Start time as a UTC datetime object.
+        expiration_time: Expiration time as a UTC datetime object. If None, an expiration time 1 hour from now will be used.
+        url_prefix: the URL prefix to sign, including protocol.
+                    For example: http://example.com/path/ for URLs under /path or http://example.com/path?param=1
+        full_path:  A full path to sign, starting with the first '/'.
+                    For example: /path/to/content.mp4
+        path_globs: a set of ','- or '!'-delimited path glob strings.
+                    For example: /tv/*!/film/* to sign paths starting with /tv/ or /film/ in any URL.
+        session_id: a unique identifier for the session
+        data: data payload to include in the token
+        headers: header name and value to include in the signed token in name=value format.  May be specified more than once.
+                    For example: [{'name': 'foo', 'value': 'bar'}, {'name': 'baz', 'value': 'qux'}]
+        ip_ranges: A list of comma separated ip ranges. Both IPv4 and IPv6 ranges are acceptable.
+                    For example: b"203.0.113.0/24,2001:db8:4a7f:a732/64"
```

Comment:
ip_ranges is pytyped to be str, so you don't want the `b` prefix in the example.

Reply:
Got it! Many thanks for reporting it.

Reply:
Update the code now.